### ISSUE TITLE
Refactored flashloader can't normalize data

### ISSUE THREAD

**Issue by zain-sohail** [OpenCOMPES/sed#459]

**Comment by rettigl** [comment]
This is somehow related to bam correction. In the timed_dataframe, delayStage positions are not the same as in the dataframe after bam correction, if parquets are created with the new flash loader. Currently, no idea why.

**Comment by rettigl** [comment]
A major difference between the old and new loader is the dask index: It is continous in the old loader, and very sparse in the new loader:
Old: 
![grafik](https://github.com/OpenCOMPES/sed/assets/53396064/e65d1c9d-3afa-4d75-b325-3023431ac629)
![grafik](https://github.com/OpenCOMPES/sed/assets/53396064/5d50a74f-4860-44d4-acb1-948de93ed993)
New: 
![grafik](https://github.com/OpenCOMPES/sed/assets/53396064/8d7e6768-30b0-41f2-adb0-27dfd27577de)
![grafik](https://github.com/OpenCOMPES/sed/assets/53396064/f7f8c703-7e88-4caa-85a0-0203f6bd2a0c)

**Comment by rettigl** [comment]
It's related to the bam correction and the mean offsetting:
Old:
![grafik](https://github.com/OpenCOMPES/sed/assets/53396064/f63f470f-f74d-4f8c-9fea-93a8d7e7e67d)
New:
![grafik](https://github.com/OpenCOMPES/sed/assets/53396064/cb17cde3-2cbc-4698-bab8-50015472d4ce)
 
Old:
![grafik](https://github.com/OpenCOMPES/sed/assets/53396064/dce1047f-dba9-44e3-95d8-9daca41379ec)
New:
![grafik](https://github.com/OpenCOMPES/sed/assets/53396064/ad2bb8f2-4eaf-43d1-9290-22c83701c033)

It seems there are a lot of empty entries in the timed_dataframe, where actually not even a photon pulse was present.

**Comment by zain-sohail** [comment]
> The computation time with the parquet generated with the new loader also is >2 as long.

I think the slowdown is because of this section being offloaded to dask
https://github.com/OpenCOMPES/sed/blob/86978c08be702f550ae10c04be1357cc012ebcf0/sed/loader/flash/loader.py#L610-L614

> It seems there are a lot of empty entries in the timed_dataframe, where actually not even a photon pulse was present.

This was requested feature by Davide (#307) that the timed dataframe should contain all available values from pulses and not just ones where electrons are detected.

**Comment by rettigl** [comment]
> This was requested feature by Davide (#307) that the timed dataframe should contain all available values from pulses and not just ones where electrons are detected

Indeed, but these should also only contain valid pulses, where actual shots exist

**Comment by zain-sohail** [comment]
> Indeed, but these should also only contain valid pulses, where actual shots exist

Yes, this comes from the flash data structure of storing 1000 pulses for each train.
The simple way I can think to adapt this is a hard threshold the pulseId for timed_dataframe. Because it's not nans that are stored in bam column but zeros. (In other pulse resolved channels might be zero)
@kutnyakhov any ideas?


> A major difference between the old and new loader is the dask index: It is continous in the old loader, and very sparse in the new loader: 

This is just because nan drop takes place after loading the buffer files.

**Comment by zain-sohail** [comment]
`sp._timed_dataframe = sp._timed_dataframe[sp_50826._timed_dataframe["pulseId"] < 500].ffill()` 
ffill should happen with overlap but in general, something like this would fix the issue of noramlizing data

**Comment by rettigl** [comment]
There should be a permanent fix. Probably beet to look for the first nonzero entry from the back or so. Another config entry I would avoid if possible

**Comment by zain-sohail** [comment]
> That's not the reason. Even if I disable the sector split, it's still slow.

Might be it's slow because of indexes
https://docs.dask.org/en/stable/generated/dask_expr._collection.DataFrame.set_index.html#dask_expr._collection.DataFrame.set_index
But I will think of a solution that allows the correct electron and timed dataframes and also be fast.

**Comment by rettigl** [comment]
If directly loading the parquet files, first thing to notice is that the new loader produces ~6.5 times more entries compared to the old one, and also the number of entries and remaining NaN is much different:
This is the output of df.count().compute on two directly loaded parquet dataframes (single file):
Old:
```
trainId                      423643
pulseId                      423643
electronId                   423643
timeStamp                    423643
dldPosX                      423643
dldPosY                      423643
dldTimeSteps                 423643
cryoTemperature              423643
dldTimeBinSize               423643
extractorCurrent             423643
extractorVoltage             423643
sampleBias                   423643
sampleTemperature            423643
tofVoltage                   423643
pulserSignAdc                  9184
monochromatorPhotonEnergy      4571
gmdBda                        44019
bam                          423341
delayStage                    43846
dldSectorID                  423643
dtype: int64
```
New:
```
trainId                      2782954
pulseId                      2782954
electronId                   2782954
dldPosX                       423643
dldPosY                       423643
dldTimeSteps                  423643
pulserSignAdc                  25630
bam                          2563000
timeStamp                       2563
monochromatorPhotonEnergy         27
gmdBda                           264
delayStage                       263
sampleBias                      2563
tofVoltage                      2563
extractorVoltage                2563
extractorCurrent                2563
cryoTemperature                 2563
sampleTemperature               2563
dldTimeBinSize                  2563
dtype: int64
```

Clearly, the performance with the old files is much superior, also their storage size is a fac.tor 2 or so smaller. This should be changed again.

**Comment by rettigl** [comment]
Regarding the timed_dataframe, does it really make a difference if this is computed on the pulse rather than the train level? This would make things super much simple and more performant. I think it's worth checking if that makes a practical difference, if you e.g. take the mean of a train and normalize by this.

**Comment by zain-sohail** [comment]
> If directly loading the parquet files, first thing to notice is that the new loader produces ~6.5 times more entries compared to the old one, and also the number of entries and remaining NaN is much different: This is the output of df.count().compute on two directly loaded parquet dataframes (single file): Old:

Yes this is simply because we are not dropping nans from electron columns before saving the files (how it was done before):
https://github.com/OpenCOMPES/sed/blob/86978c08be702f550ae10c04be1357cc012ebcf0/sed/loader/flash/loader.py#L610-L614
If we do that, the dataframe becomes much smaller.
This doesn't work with if we want to keep the pulse data even without electrons. 
One solution could be to save a df and df_timed parquet and then load both per file. Technically it's also possible to save a parquet for each run and just divide the 'per file' into row groups.

> Clearly, the performance with the old files is much superior, also their storage size is a fac.tor 2 or so smaller. This should be changed again.

I can of course bring back the old behavior (saving after dropping nans) so we can run tests to see if that is the performance bottleneck.
File size is not the biggest issue because parquet files are quite small.

> Regarding the timed_dataframe, does it really make a difference if this is computed on the pulse rather than the train level? This would make things super much simple and more performant. I think it's worth checking if that makes a practical difference, if you e.g. take the mean of a train and normalize by this.

I thought we  just had a discussion today about monochromator and using the pulse channel for it instead of train. Perhaps I misunderstood but I can't shed light on whether it's useful or not.

@balerion if you are still using sed, could you explain your usecase?

**Comment by zain-sohail** [comment]
as expected, from my tests, binning performance stays same if i remove nans before saving (index is monotonic then like in main branch version. You can test from branch fix-459

**Comment by rettigl** [comment]
Something is still different. While the speed is back to normal, the result looks quite different:
Old:
![image](https://github.com/OpenCOMPES/sed/assets/53396064/7c09d152-0189-47d6-b418-f0a5c068af9a)
![image](https://github.com/OpenCOMPES/sed/assets/53396064/2c726bea-8cdc-4586-b91b-6982b8572cec)
![image](https://github.com/OpenCOMPES/sed/assets/53396064/e39802dd-6bf8-47f0-b873-71edfe602f71)
New:
![image](https://github.com/OpenCOMPES/sed/assets/53396064/4625a5a1-4f75-4cf3-b6ed-2dfc49df1579)
![image](https://github.com/OpenCOMPES/sed/assets/53396064/d18334d8-2c94-4a22-bd80-3fcc7deba517)
![image](https://github.com/OpenCOMPES/sed/assets/53396064/68a4d878-c436-424d-89a9-5d9a934175de)
Fixed:
![image](https://github.com/OpenCOMPES/sed/assets/53396064/e079fd5e-dc9f-4e60-9c39-4ea69097113d)
![image](https://github.com/OpenCOMPES/sed/assets/53396064/2e0f0f27-9e0a-493b-90ee-e8d9651f164a)
![image](https://github.com/OpenCOMPES/sed/assets/53396064/c2eb5bae-58e5-4759-b49f-ec2baba44fed)

It appears to attach the wrong delay stage values (?)

**Comment by rettigl** [comment]
Counts for the same file:
Old:
![image](https://github.com/OpenCOMPES/sed/assets/53396064/7c450ec1-ad18-474b-9384-20ad5078561c)
New:
![image](https://github.com/OpenCOMPES/sed/assets/53396064/671a6e01-61d0-4b80-8cc7-0330a26b921f)
Fixed:
![image](https://github.com/OpenCOMPES/sed/assets/53396064/c298cbdd-b8c8-47a8-911c-ac9bba08ae2a)

There are 2567 trains in this file, and 257 different unique delay stage position values

**Comment by zain-sohail** [comment]
closed by #469
